db/db_impl/db_impl_write.cc

```diff
@@ -686,9 +686,10 @@ Status DBImpl::WriteImplWALOnly(
     // TODO(myabandeh): Make preliminary checks thread-safe so we could do them
     // without paying the cost of obtaining the mutex.
     if (status.ok()) {
-      InstrumentedMutexLock l(&mutex_);
+      mutex_.Lock();
       bool need_log_sync = false;
       status = PreprocessWrite(write_options, &need_log_sync, &write_context);
+      mutex_.Unlock();
```

Comment:
I'm worried if this might expose some race conditions. For example, if PreprocessWrite() returns an error status due to bg_error_, but it gets cleared before the call to WriteStatusCheck(). Maybe we can define a WriteStatusCheckLocked() and make WriteStatusCheck() a wrapper around it. PreprocessWrite() can directly call WriteStatusCheckLocked().

Reply:
I see, maybe WriteStatusCheckLocked() is a better solution for this.